### PR TITLE
Basic API

### PR DESCRIPTION
**Pull request by wildmolasses**

### CODE REVIEW

File: api/create-table.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+const connection = new AppDAO('./database.sqlite3')
+
+
+const sql = `
+    CREATE TABLE IF NOT EXISTS block_address (
+      block_number INTEGER NOT NULL,
+      tx_index INTEGER,
+      trace_id INTEGER,
+      address VARCHAR(42) NOT NULL,
+      reason VARCHAR(255) NOT NULL
```
Comment (tjayrush) [inline api/create-table.js]: Give me another field here called 'proof'

File: api/README.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+Instructions:
+
+```
+cd api
+npm install
+npm run dev
+```
+
+The API expects an array of JSON. Here is an example of the POST request data the application expects.
+
```
Comment (tjayrush) [inline api/README.md]: Please, please, please, please, please let me send http://localhost:8000/api/v1/bn=123&txid=456&traceid=12&address=0x12345&reason=whatever&proof=nothing.

(or the equivelant POST)

So much easier. So much less data over the wire.

Comment (wildmolasses) [reply api/README.md]: @tjayrush OK, but I want you to be able to batch inserts. We can do it without large JSON requests, but to get started, this was easiest.

File: api/create-table.js (hunk 1 above)
Comment (wildmolasses) [reply api/create-table.js]: Easily. Can you try running the app locally and writing data to it?

File: api/README.md (hunk 2 above)
Comment (tjayrush) [reply api/README.md]: It works.

**Comment by tjayrush** [comment]
This works for now. Some notes, just FYI.

1) both txIndex and traceId can be empty
2) is it normal to have txIndex in the POST and tx_index in the database? Is that just standard naming? I'm asking because I want to be normal, but it seems inconsistent. Same question for traceId and trace_id. Do you name them that way to make it clear which field you're talking about?
3) dups are on blockNum,txIndex,traceId
4) probably don't need to deliver address or reason (or proof) by default
5) add a proof field (32 bytes + 2 bytes for multi-hash + 1 byte for '0x', so 36 bytes?

I changed my mind about above comment. JSON is fine. No problem at all. It works perfectly with existing C++ code.

File: api/create-table.js (hunk 1 above)
Comment (wildmolasses) [reply api/create-table.js]: Hey Jay, re: proof field, do you have a specific hashing algorithm in mind? I believe that the length of the multi-hash is dependent on the length of your hashed proof.

Comment (wildmolasses) [reply api/create-table.js]: Nevermind, I see in your comments that you're thinking 32 bytes.

**Comment by wildmolasses** [comment]
> This works for now. Some notes, just FYI.
> 
> 1. both txIndex and traceId can be empty

Sounds good: create-table.js creates a table that allows null values for txIndex and traceId. (https://github.com/wildmolasses/trueblocks/blob/7cbc69bd711ed99489eb0837fd651dfe3c5aaf47/api/create-table.js#L8)

> 2. is it normal to have txIndex in the POST and tx_index in the database? Is that just standard naming? I'm asking because I want to be normal, but it seems inconsistent. Same question for traceId and trace_id. Do you name them that way to make it clear which field you're talking about?

My logic is as follows: camelcase is traditional in javascript, and underscores are traditional in sql (case sensitivity inconsistent between different sql implementations). That being said, if it's confusing, I'd be happy to change it.

> 3. dups are on blockNum,txIndex,traceId

And address?

> 4. probably don't need to deliver address or reason (or proof) by default

Re: excluding address, are we talking about the case where people query by address, therefore not needing address printed as a column? Besides this case, address seems necessary.

> 5. add a proof field (32 bytes + 2 bytes for multi-hash + 1 byte for '0x', so 36 bytes?

Will do. I'll be happy when we move from varchar to binary storage. Meanwhile, I'll make this a VARCHAR(36).

> I changed my mind about above comment. JSON is fine. No problem at all. It works perfectly with existing C++ code.

Let's support other formats too!

**Comment by wildmolasses** [comment]
Some thoughts of mine:

- This data depends on when it was retrieved and what the latest block was at that time. Perhaps that should be represented in the query result.
- Optimize data storage where possible (varchar -> binary)
- What routes and query parameters belong on API? We want to query by address. What else? Block window (e.g. between blocks 5m and 6m)?
- In production, POST requests should be authenticated.
- How to manage the insertion process? What if Jay posts duplicates, or what if his POST times out, fails?

**Comment by tjayrush** [comment]
> This data depends on when it was retrieved and what the latest block was at that time. Perhaps that should be represented in the query result.

I'm not sure I understand what you mean by this.

> Optimize data storage where possible (varchar -> binary)

Save half the space of storing an address

> What routes and query parameters belong on API? We want to query by address. What else? Block window (e.g. between blocks 5m and 6m)?

I feel like we want to guard against doing too much for the user. We want to promote the idea that running a node is beneficial. If we offer too much from this API, the user won't run their own node (we can debate whether that is a good thing or a bad thing.) For now -- we only query by address (perhaps within a block range). The delivered data is used to extract the remaining data directly from the node.

> In production, POST requests should be authenticated.

If we feed data through the API. If we can figure out a way to run a node on Tokenomics, then it could write directly into the back end, but as you point out, this has downsides not the least of which is portability.

> How to manage the insertion process? What if Jay posts duplicates, or what if his POST times out, fails?

Dups need to be removed. I'd kind of like this process to be on the tokenomics server, but that means running a tracing node on Digital Ocean. Pain all around.

**Pull request merged**